### ISSUE TITLE
[psyir] Check on whether dimensioning symbol is local does not work when symbol is in an ancestor scope.

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#1089]
For the following Fortran:

    module my_mod
      integer :: dim
    contains
      subroutine init()
        dim = 10
      end subroutine
      subroutine work()
        real, dimension(dim) :: my_array
      end subroutine
    end module

Then PSyclone complains:

    If a local datasymbol is used as part of a dimension declaration then it should be a constant, but 'dim' is not.

The problem is in `ArrayType._validate_shape` where the Symbol being referred to in the shape specification is looked up. It is found to have a 'local' interface and not be a constant. Both of which are correct. However, no check is made on which SymbolTable the symbol is in.

**Comment by arporter** [comment]
I've just discovered #948 which is linked to an existing, xfailing test. Strangely, that issue says the new checks are 'commented out'? Does this ring any bells @rupertford?

**Comment by rupertford** [comment]
> I've just discovered #948 which is linked to an existing, xfailing test. Strangely, that issue says the new checks are 'commented out'? Does this ring any bells @rupertford?

I guess it was out-of-date. Thanks for closing.

**Comment by arporter** [comment]
I can't see any way to fix this other than to remove the check:
https://code.metoffice.gov.uk/trac/lfric/browser/LFRic/trunk/gungho/source/kernel/advection/poly2d_flux_kernel_mod.F90
The reason is that we need to know which symbol table contains the referenced symbol. Although `Symbol` has the `find_symbol_table` method, this requires a PSyIR node from which to search. The `_validate_shape` method is within the `ArrayType` class and therefore the only PSyIR node available is the `Reference` to the symbol in the shape specification. Since we are normally constructing an array reference at this point, this `Reference` doesn't yet have a parent since we need the type definition before we can create the `ArrayReference`.
Can @rupertford or @sergisiso think of any solution? What was the motivation for adding this check originally?

**Comment by arporter** [comment]
We could move the check to the backend as, at that point, we know we must have a fully-formed PSyIR tree.

**Comment by sergisiso** [comment]
The problem is this is not the main tree it is a purpose made tree inside the shape list. We have the associations:

PSyIR Tree Node -> SymbolTable -> DataSymbol -> DataType -> shape list -> Reference -> symbol

So this Reference will never have a scope with the current implementation?

**Comment by sergisiso** [comment]
Maybe should manually point (somehow) that the scope of the shape list PSyIR elements is the same SymbolTable that contain the DataSymbol they belong? When the shape list PSyIR is constructed?

**Comment by arporter** [comment]
Unless I've missed something, I think the problem is that we construct the shape list *before* we create the symbol to which it belongs (because we have to make the type first). We'd need to alter the `DataSymbol` constructor so that `datatype` was optional.
Rather than re-engineer the PSyIR classes, I think it would be more natural to do this check in e.g. `psyir.backends.fortran.FortranWriter.gen_decls` since that knows which symbol table it is dealing with.

**Comment by rupertford** [comment]
Isn't this an example of a check that the PSyIR is valid/consistent. There will probably be many other checks we could/should perform. PSyIR checks are independent of the backend, so I don't think that is the right place to put them, unless we ensure all backends call the same generic checks via some subclassing?

**Comment by arporter** [comment]
Yes, that's what this is. Do we need another 'validation' visitor?

**Comment by rupertford** [comment]
A visitor sounds good to me to traverse the tree. I'm not sure whether it's more maintainable for the checks to go with the node or visitor though.

**Comment by sergisiso** [comment]
This may be needed but I first would explore that we are not missing to describe some info that could help without the validation pass. For instance:
Should this symbol really have a LocalInterface? An alternative is that all symbols declared in a Container should have a distinct interface (for lack of a better name GlobalInterface, while imported symbols should be renamed to ExternalInterface). Would this solve the issue here?

**Comment by arporter** [comment]
That does sound a better way of thinking about it - clearly, having a symbol described as 'local' to a Container isn't very useful as there's no executable code that it is 'local' to. Any code must obviously be in a routine within the Container, at which point that symbol is not 'local'.

**Comment by rupertford** [comment]
I'm not sure I agree with the idea that a symbol within a container can't be local. Within a particular scope a symbol is either visible externally or it is not. Variables and procedures can be declared local to a module, or be declared shared. The point is that something is local to its scope, so for nested scopes something might be local in an outer scope but global/shared/visible within an inner scope.

**Comment by arporter** [comment]
While that's true @rupertford, that's not what we have implemented at the moment. The 'interface' of a symbol is a property of that symbol whereas whether something is global/shared/local actually depends on which symbol table is being queried. Currently, `symbol.is_local` simply queries the `interface` property of the symbol object. We could change this so that an (optional) symbol table could be supplied as part of the query. However, that wouldn't solve the immediate problem as we don't have access to a symbol table while constructing a data type.

**Comment by rupertford** [comment]
Whilst this is not useful for you, perhaps we could have a `SymbolTable` method `is_local(symbol)` or equivalent?

**Comment by rupertford** [comment]
I do wonder though, whether this problem is due to the fact that we are not really creating a datatype, we are mixing the datatype with an instance of a datatype. A datatype should just declare types and therefore should not reference symbols - or am I talking rubbish?

**Comment by sergisiso** [comment]
The Local interface doesn't mean local scope (maybe its also not the best name?). To know if something is local/global to its scope we have to look at what symbol table it is in (and its visibility to know if can be accessed externally).

The interface tell us how to declare/access/what is the lifetime of the variable. Maybe it should be named AutomaticInterface (as in the C standard), meaning that is a stack-allocated variable that will disappear at the end of the scope. The container variables (globals) do not have this same kind of Interface.

I believe this could solve the dimension symbol issue without having to access to the symbol_table, meaning that it can be done in the datatype creation (if I am not missing something).

**Comment by rupertford** [comment]
OK, I'm not understanding the difference between lifetime and scope. I'll wait to be educated at some point :-)

**Comment by sergisiso** [comment]
> That does sound a better way of thinking about it - clearly, having a symbol described as 'local' to a Container isn't very useful as there's no executable code that it is 'local' to. Any code must obviously be in a routine within the Container, at which point that symbol is not 'local'.

There is, e.g. in the parameter initialization PSyIR trees:
```
module mymod
  use other, only: a, b
  real, parameter :: c = 4
  real, parameter :: value = a + 3 * b - c
```
In this case c is local to the module but I would remove Local as an interface and just let implicitly be local whatever exist in the .scope.symbotalbe.

**Comment by hiker** [comment]
I had a bit of a look, it appears that the distinction between a 'routine symbol' (defined somewhere in the routine) and a `module symbol` (a symbol defined in the outside module) could be done in `fparser2._process_decln`: if the `parent` is a container, then the declaration is in the outer module, otherwise it appears to be local.

I have the feeling that this issue is better with @sergisiso , e.g. I am not sure if we want to add another `is_outer_module` attribute, or change LocalInterface to something else as Sergi suggested - he's the expert here ;)

For my use case I have a simple loop to find the right symbol table, something like:
```
            node = ref    # Try to find out if the given Reference `ref` is local or in the outer module:
            while node:
                if hasattr(node, "_symbol_table"):
                    if sym.name in node.symbol_table:
                        if isinstance(node, Container):
                            # It is a variable from the module in which the
                            # current function is, so it is an external
                            # reference
                            result.append((sym, node.name))
                        # If the node is not the container, it is a local
                        # variable and does not need to be returned.
                        break
                # Otherwise keep on looking
                node = node.parent
```
That seems to work just fine, and I will add a TODO for this ticket, so it can be simplified later.
